pyodide-build/pyodide_build/out_of_tree/venv.py

```diff
@@ -0,0 +1,287 @@
+    """Write script to invoke host python into the virtualenv bin folder"""
+    import os
+
+    # Resetting PATH and VIRTUAL_ENV will temporarily restore us to the virtual
+    # environment that 'pyodide venv' was invoked in (or no virtual environment
+    # if we aren't currently in one).
+    environment_vars = []
+    for environment_var in ["VIRTUAL_ENV", "PATH", "PYODIDE_ROOT"]:
+        value = os.environ.get(environment_var, "''")
+        environment_vars.append(f"{environment_var}={value}")
+    environment = " ".join(environment_vars)
+
+    host_python_path = venv_bin / "python3.10-host"
+    host_python_path.write_text(
+        dedent(
+            f"""
+            #!/bin/sh
+            {environment} exec {sys.executable} $@
+            """
+        )
+    )
+    host_python_path.chmod(0o777)
+
+
+def create_pyodide_script(venv_bin: Path) -> None:
+    """Write pyodide cli script into the virtualenv bin folder"""
+    import os
+
+    # The pyodide cli must be invoked outside of the virtual env. In order to
+    # ensure this, we have to restore VIRTUAL_ENV, PATH, and PYODIDE_ROOT to
+    # their current values before invoking it..
+    # In case we are inside an xbuildenv, make sure to make the path relative to
+    # this file and not relative to PYODIDE_ROOT.
+    pyodide_build_path = str(Path(__file__).parents[2])
+
+    # Resetting PATH and VIRTUAL_ENV will temporarily restore us to the virtual
+    # environment that 'pyodide venv' was invoked in (or no virtual environment
+    # if we aren't currently in one). To be extra careful, we set PYTHON_PATH too.
+    environment_vars = [f"PYTHONPATH={pyodide_build_path}"]
+    for environment_var in ["VIRTUAL_ENV", "PATH", "PYODIDE_ROOT"]:
+        value = os.environ.get(environment_var, "''")
+        environment_vars.append(f"{environment_var}={value}")
+    environment = " ".join(environment_vars)
+
+    pyodide_path = venv_bin / "pyodide"
+    pyodide_path.write_text(
+        dedent(
+            f"""
+            #!/bin/sh
+            {environment} exec {sys.executable} -m pyodide_build.out_of_tree $@
+            """
+        )
+    )
+    pyodide_path.chmod(0o777)
+
+
+def install_stdlib(venv_bin: Path) -> None:
+    """Install micropip and all unvendored stdlib modules"""
+    # Micropip we can install with pip!
+    toload = ["micropip"]
+    result = subprocess.run(
+        [venv_bin / "pip", "install", *toload],
+        capture_output=True,
+        encoding="utf8",
+    )
+    check_result(result, "ERROR: failed to invoke pip")
+
+    # Other stuff we need to load with loadPackage
+    # TODO: Also load all shared libs.
+    result = subprocess.run(
+        [
+            venv_bin / "python",
+            "-c",
+            dedent(
+                """
+                from _pyodide._importhook import UNVENDORED_STDLIBS_AND_TEST;
+                from pyodide_js import loadPackage;
+                loadPackage(UNVENDORED_STDLIBS_AND_TEST);
```

Comment:
Do we need to load test?

Reply:
I guess we don't need to but I don't see why not do so. It's around in the native Python.

Reply:
Oh, I was misunderstanding the code here. I thought that this needs to be called every time, but it looks like it only happens onetime at the virtualenv initialization step. Nice!